vmm/src/api/http/http_endpoint.rs

```diff
@@ -50,6 +52,162 @@ use crate::config::RestoreConfig;
+        /// device attach. In these cases, only one single config object is
+        /// handled by the API call.
+        fn expected_num_fds(&self) -> Option<usize>;
+
+        /// Returns any FDs provided in the HTTP body.
+        ///
+        /// They will always be invalid. Used for logging.
+        fn fds_from_http_body(&self) -> Option<&[i32]>;
+
+        /// Updates the internal fds field to a new value.
+        fn set_fds(&mut self, fds: Option<Vec<i32>>);
+    }
+
+    mod config_with_fds_impls {
+        use super::ConfigWithFDs;
+        use crate::config::RestoredNetConfig;
+        use crate::vm_config::NetConfig;
+
+        impl ConfigWithFDs for NetConfig {
+            fn id(&self) -> Option<&str> {
+                self.id.as_deref()
+            }
+
+            fn expected_num_fds(&self) -> Option<usize> {
+                None
+            }
+
+            fn fds_from_http_body(&self) -> Option<&[i32]> {
+                self.fds.as_deref()
+            }
+
+            fn set_fds(&mut self, fds: Option<Vec<i32>>) {
+                self.fds = fds;
+            }
+        }
+
+        impl ConfigWithFDs for RestoredNetConfig {
+            fn id(&self) -> Option<&str> {
+                Some(self.id.as_str())
+            }
+
+            fn expected_num_fds(&self) -> Option<usize> {
+                Some(self.num_fds)
+            }
+
+            fn fds_from_http_body(&self) -> Option<&[i32]> {
+                self.fds.as_deref()
+            }
+
+            fn set_fds(&mut self, fds: Option<Vec<i32>>) {
+                self.fds = fds;
+            }
+        }
+    }
+
+    /// Applies FDs to configs for their corresponding devices, as part of the special
+    /// handling for devices backed by externally provided FDs.
+    ///
+    /// The FDs (via `files`) **must** be provided in the exact order matching the
+    /// config struct they belong to.
+    ///
+    /// See [module description] for more info.
+    ///
+    /// [module description]: self
+    pub fn apply_new_fds_to_cfg<T: ConfigWithFDs>(
+        // List of new files (well, actually FDs) that back up the device(s) of the given config(s).
+        files: Vec<File>,
+        // List of network configurations where each network can have up to `n` FDs.
+        cfgs: &mut [&mut T],
+    ) -> Result<(), HttpError> {
+        // Check: Only one single cfg wants to take ownership of all `files`.
+        {
+            let wants_ownership_of_all_files =
+                cfgs.iter().any(|cfg| cfg.expected_num_fds().is_none());
+            if wants_ownership_of_all_files {
+                assert_eq!(
+                    cfgs.len(),
+                    1,
+                    "should have exactly one cfg when expected_num_fds() returns None"
+                );
+            }
+        }
+
+        let expected_fds_total: usize = cfgs
+            .iter()
+            // Note: if this is None, `cfgs` must have a length of `1`!
```

Comment:
Why not a different function, rather than having one function handle two different input cases, then?

Reply:
I can try to give it a shot. I'm concerned that we will end up with some code duplication

Reply:
My instinct is that duplication might be the lesser of two evils here, but let's see what can be factored out.